### PR TITLE
use requests when it is installed

### LINKED ISSUE #16716: read_csv from HTTPs + basic-auth + custom port throws an error (urlopen error)
#### Code Sample, a copy-pastable example if possible

```python
import pandas as pd
df = pd.read_csv('https://user-name:pwd@my.fqdn.com:8080/get_content.csv')
# fails with urlopen error [Errno 11003] getaddrinfo failed Python2.7 as well as Python 3.x
```
#### Problem description

HTTPS basic auth is very common. This URL format works in Excel, other text editors, etc.  This url works in `requests ` library.  It seems like the scenario doesnt work because the underlying `urlopen` doesnt work
```python
# I think it fails because the underlying code here fails in pandas/io/common.py:
from urllib2 import urlopen as _urlopen
b = _urlopen('https://un-name:pwd@my.fqdn.com:8080/get_content.csv')
# fails with urlopen error [Errno 11003] getaddrinfo failed
```

see [stackoverflow issue urllib basic auth](https://stackoverflow.com/questions/635113/python-urllib2-basic-http-authentication-and-tr-im)
Only way to overcome this is using `requests` + `StringIO` ?

#### Expected Output
be able to get a CSV loaded dataframe

#### Output of ``pd.show_versions()``

<details>
python: 2.7.13.final.0
python-bits: 64
pandas: 0.20.1
</details>

### LINKED ISSUE #28825: UnsupportedOperation 'seek' when loading excel files from urllib3.response.HTTPResponse object
0.25.1 seems to have reintroduced #20434 

**Update:** not the case, but the same problem happens when feeding `excel_read()` with an [urllib3.response.HTTPResponse](https://urllib3.readthedocs.io/en/latest/reference/#urllib3.response.HTTPResponse) object. Code sample updated.

#### Code Sample, a copy-pastable example if possible
```
import pandas
import requests

url = "https://raw.github.com/pandas-dev/pandas/master/pandas/tests/io/data/test1.xls"
r = requests.get(url, stream=True)
df = pandas.read_excel(r.raw)
print(df)
```

#### Problem description
Using `0.25.1` I get the same problem and output as described in #20434 with the code above. 

#### Expected Output
See #20434

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : None

pandas           : 0.25.1
numpy            : 1.17.2
pytz             : 2019.3
dateutil         : 2.8.0
pip              : 19.2.3
setuptools       : 41.4.0
Cython           : None
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 2.10.3
IPython          : None
pandas_datareader: None
bs4              : None
bottleneck       : None
fastparquet      : None
gcsfs            : None
lxml.etree       : None
matplotlib       : None
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pytables         : None
s3fs             : None
scipy            : None
sqlalchemy       : None
tables           : None
xarray           : None
xlrd             : 1.2.0
xlwt             : None
xlsxwriter       : None

</details>

### PR DESCRIPTION
**Pull request by ocefpaf**
closes #16716
closes #28825
closes #28826
solves https://github.com/pandas-dev/pandas/pull/16910

- [ ] tests added / passed
- [X] passes `black pandas`
- [X] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

Continuation of https://github.com/pandas-dev/pandas/pull/21504 and https://github.com/pandas-dev/pandas/pull/17087

This PR is not ready but I'd appreciate any early comment/feedback.

### CODE REVIEW

File: pandas/io/common.py [hunk 1]
```diff
@@ -183,6 +183,37 @@ def is_gcs_url(url) -> bool:
     return parse_url(url).scheme in ["gcs", "gs"]
 
 
+def _urlopen(*args, **kwargs):
```
Comment (ocefpaf) [inline pandas/io/common.py]: Should we just make a single `urlopen` here?

File: pandas/io/excel/_base.py [hunk 2]
```diff
@@ -336,10 +336,10 @@ def read_excel(
 
 
 class _BaseExcelReader(metaclass=abc.ABCMeta):
-    def __init__(self, filepath_or_buffer):
+    def __init__(self, filepath_or_buffer, session=None):
```
Comment (ocefpaf) [inline pandas/io/excel/_base.py]: I'm a bit lost on how to pass the `session` object to this class, right now it `defaults` to `None` b/c it does not get the value from `read_excel`.

File: pandas/io/excel/_base.py [hunk 3]
```diff
@@ -20,7 +20,7 @@
     _stringify_path,
     _validate_header_arg,
     get_filepath_or_buffer,
-    urlopen,
+    _urlopen,
```
Comment (jbrockmendel) [inline pandas/io/excel/_base.py]: can you use the non-private version

File: pandas/io/common.py [hunk 4]
```diff
@@ -183,13 +183,46 @@ def is_gcs_url(url) -> bool:
...
+                    "Expected a requests.sessions.Session object, "
+                    "got {!r}".format(session)
+                )
+            r = session.get(url)
+        else:
+            r = requests.get(url)
+        r.raise_for_status()
+        content = r.content
+        r.close()
+    except ImportError:
```
Comment (jbrockmendel) [inline pandas/io/common.py]: can you use pandas.compat._optional.import_optional_dependency here

**Comment by WillAyd** [comment]
Haven't read through the attached issue but this is a little more complicated than I would have hoped for. Is there a reason why you think this is in scope of pandas? Pandas can operate for the most part on items that adhered to the IOBase definition. 

I don't think dealing with the `requests.models.Response` object is within scope

File: pandas/io/common.py (hunk 1 above)
Comment (jreback) [reply pandas/io/common.py]: yes call this urlopen

File: pandas/io/common.py [hunk 5]
```diff
@@ -183,13 +183,46 @@ def is_gcs_url(url) -> bool:
     return parse_url(url).scheme in ["gcs", "gs"]
 
 
+def _urlopen(*args, **kwargs):
+    compression = None
+    content_encoding = None
+    try:
+        import requests
```
Comment (jreback) [inline pandas/io/common.py]: need to use the pandas.compat._optional.import_optional_dependency

**Comment by ocefpaf** [comment]
> IIRC pandas_datareader supports requests Sessions

I know. That is what inspired the original poster of this PR.

I'm traveling back home today and I'll address the review comments during the weekend. Thanks!

**Comment by WillAyd** [comment]
I guess I'm not clear on what Sessions are doing here. Neither of the issues that this closes reference Session objects anywhere - are you sure this closes them?

Can you add test cases?

File: pandas/io/common.py (hunk 4 above)
Comment (ocefpaf) [reply pandas/io/common.py]: Not sure if I got it right.

File: pandas/io/common.py (hunk 1 above)
Comment (ocefpaf) [reply pandas/io/common.py]: Done.

**Review by ocefpaf — COMMENTED** [review]
> I guess I'm not clear on what Sessions are doing here. Neither of the issues that this closes reference Session objects anywhere - are you sure this closes them?

The original issue was https://github.com/pandas-dev/pandas/issues/16716 by @skynss and the original PR used session. See https://github.com/pandas-dev/pandas/pull/17087

Allow us to pass user/password, for example, for password protected servers.
One can also use a local `.netrc` file. Let's remove the session for now to see if we can make progress.

> Can you add test cases?

Sure. Writing those, just wanted to be sure this will get a green signal before investing more time on it.

File: pandas/io/common.py [hunk 6]
```diff
@@ -184,13 +185,25 @@ def is_gcs_url(url) -> bool:
...
+        content = r.content
+        r.close()
+    except ImportError:
+        import urllib.request
 
-    return urllib.request.urlopen(*args, **kwargs)
+        r = urllib.request.urlopen(*args, **kwargs)
+        content = r.read()
+        content_encoding = r.headers.get("Content-Encoding", None)
+    if content_encoding == "gzip":
```
Comment (TomAugspurger) [inline pandas/io/common.py]: I think this also needs to be under the `except ImportError`? From what I can tell, requests `.content` will automatically decode gzip: https://requests.readthedocs.io/en/master/user/quickstart/#binary-response-content

**Review by TomAugspurger — COMMENTED** [review]
FWIW, we've been having multiple conversations about how users pass through kwargs from a read like `pd.read_csv(...)` to the "file open" call. Sometimes it's just the builtin `open(path)`, but in this case it might be `requests.get(...)` At some point, I'd like to unify these into a `backend_kwargs` or `open_kwargs` dict that gets passed through (but that's for a separate issue. I'll try to open one).

**Comment by WillAyd** [comment]
@ocefpaf still active? Can you address comments and check CI failures?

**Comment by ocefpaf** [comment]
I'll probably have some time for this during the end of the year break. If someone wants to pick it up from here before that please feel free to do so.

**Comment by jreback** [comment]
@ocefpaf if you can merge master and see if you can get passing would be great.

**Comment by MarcoGorelli** [comment]
Hi @ocefpaf  - sorry to chase you up, just wanted to ask whether you're still working on this :)
Seems that there's some test failures, but I can't check them as it says "build not found"

**Comment by WillAyd** [comment]
Closing as I think stale, but certainly ping if you'd like to continue

**Pull request closed without merging**